File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260921-204934-85557/workdir/src/main/java/surveycrawler/CrawlerCommand.java
package surveycrawler;

public abstract class CrawlerCommand {

    // instance creation

    public static CrawlerCommand commandFor(char aCharacter) {
        CrawlerCommand[] commandPrototypes = commandPrototypes();
        for (int i = 0; i < commandPrototypes.length; i++) {
            if (commandPrototypes[i].isFor(aCharacter)) {
                return commandPrototypes[i].forCharacter(aCharacter);
            }
        }
        return new InvalidCommand();
    }

    private static CrawlerCommand[] commandPrototypes() {
        return new CrawlerCommand[] {
            new AdvanceCommand(),
            new RetreatCommand(),
            new TurnClockwiseCommand(),
            new TurnCounterClockwiseCommand(),
            new StartGuardedRunCommand(),
            new EndGuardedRunCommand(),
            new RepeatLastCommand('0')
        };
    }

    protected CrawlerCommand forCharacter(char aCharacter) {
        return this;
    }

    // testing

    public abstract boolean isFor(char aCharacter);

    // executing

    public abstract void executeOn(SurveyCrawler aCrawler);

    /* Only the commands the crawler can repeat understand how to do it, the rest of them are
       an invalid command when a digit follows them */
    public void repeatTimesOn(int aNumberOfTimes, SurveyCrawler aCrawler) {
        aCrawler.signalInvalidCommand();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-202806-59410/workdir/src/surveycrawler/crawler_command.py
class CrawlerCommand:

    # instance creation

    @classmethod
    def for_character(cls, a_character):
        from surveycrawler.invalid_command import InvalidCommand

        for a_crawler_command_class in cls.__subclasses__():
            if a_crawler_command_class.is_for(a_character):
                return a_crawler_command_class()
        return InvalidCommand()

    @classmethod
    def no_command(cls):
        from surveycrawler.invalid_command import InvalidCommand

        return InvalidCommand()

    # command character

    @classmethod
    def command_character(cls):
        raise NotImplementedError()

    # testing

    @classmethod
    def is_for(cls, a_character):
        return cls.command_character() == a_character

    # repeating

    @classmethod
    def number_of_repetitions_of(cls, a_digit_character):
        return int(a_digit_character) + 2

    def repeat(self, a_number_of_repetitions, a_crawler):
        for _ in range(a_number_of_repetitions):
            self.execute(a_crawler)

    # executing

    def execute(self, a_crawler):
        raise NotImplementedError()
